File: java/src/main/java/vsdk/toolkit/galerkin/processing/visitors/ProjectedAreaAccumulatorVisitor.java
package vsdk.toolkit.galerkin.processing.visitors;

import vsdk.toolkit.galerkin.GalerkinElement;
import vsdk.toolkit.galerkin.GalerkinState;
import vsdk.toolkit.galerkin.processing.ClusterTraversalStrategy;

public final class ProjectedAreaAccumulatorVisitor implements ClusterLeafVisitor {
    private double totalProjectedArea;

    public ProjectedAreaAccumulatorVisitor() {
        totalProjectedArea = 0.0;
    }

    @Override
    public void visit(
        GalerkinElement galerkinElement,
        GalerkinState galerkinState) {
        totalProjectedArea += ClusterTraversalStrategy.surfaceProjectedAreaToSamplePoint(galerkinElement);
    }

    public double getTotalProjectedArea() {
        return totalProjectedArea;
    }
}

File: typescript/src/main/vsdk/toolkit/galerkin/processing/visitors/ProjectedAreaAccumulatorVisitor.ts
import { GalerkinElement } from "../../GalerkinElement";
import { GalerkinState } from "../../GalerkinState";
import { ClusterTraversalStrategy } from "../ClusterTraversalStrategy";
import { ClusterLeafVisitor } from "./ClusterLeafVisitor";

export class ProjectedAreaAccumulatorVisitor implements ClusterLeafVisitor {
  private totalProjectedArea: number;

  public constructor() {
    this.totalProjectedArea = 0.0;
  }

  public visit(
    galerkinElement: GalerkinElement,
    _galerkinState: GalerkinState,
  ): void {
    this.totalProjectedArea += ClusterTraversalStrategy.surfaceProjectedAreaToSamplePoint(galerkinElement);
  }

  public getTotalProjectedArea(): number {
    return this.totalProjectedArea;
  }
}
